design-proposals/descheduler-support.md

```diff
@@ -0,0 +1,106 @@
+# Overview
+Deschedulers are Kubernetes components that act as load balancers for compute nodes.
+Their goal is to spread the workload in the cluster.
+
+The descheduler's way of moving Pods around is using the [eviction API](https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/).
+Moving a regular pods is relatively simple, since they are supposed to be ephemeral.
+On the contrary, moving a `virt-launcher` pod that contains a running VM is more challenging, since VMs are stateful.
+
+Currently, there is an issue for deschedulers to work alongside KubeVirt because of the unique behavior of `virt-launcher` pods:
+
+From the descheduler's point of view, `virt-launcher` pods fail to be evicted, but they actually do migrate to another node in the background.
+The descheduler notes the failure to evict the `virt-launcher` pod and keeps trying to evict other pods, typically resulting in it attempting to evict substantially all pods from the node.
+In other words, the way KubeVirt handles eviction requests causes the descheduler to make wrong decisions and take wrong actions that could destabilize the cluster.
+
+This document proposes the addition of functionality intended to support the execution of KubeVirt in clusters that also have a descheduler.
+
+## Participating Components
+The following sections will describe how eviction requests are currently handled by KubeVirt.
+
+### Pod Eviction Webhook
+There is a [validating webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/) that intercepts all eviction requests created in the cluster.
+The webhook approves all eviction requests for pods that are not `virt-launcher` pods.
+
+The webhook takes action, based on the VMI's eviction strategy (defined on the VMI with a fallback to a cluster-wide definition):
+
+| Migration Strategy    | VMI is migrateable                    | VMI is not migrateable | Remarks                               |
+|-----------------------|---------------------------------------|------------------------|---------------------------------------|
+| None                  | N/A                                   | N/A                    | Always approve                        |
+| LiveMigrate           | Mark the VMI for eviction and approve | Deny                   |                                       |
+| LiveMigrateIfPossible | Mark the VMI for eviction and approve | Approve                |                                       |
+| External              | N/A                                   | N/A                    | Mark the VMI for eviction and approve |
+
+
+### Disruption Budget Controller
+`virt-launcher` pods have another layer of protection against evictions using [PDBs](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-budgets).
```

Comment:
I don't see this part in the design proposal anymore. Anyway I don't think they can be described as 2 layers of protection, they have different roles, the webhook marks the VMI so that it can be migrated by the evacuation controller while the PDB just blocks the eviction request from k8s.

Reply:
Yes, my question is irrelevant to this proposal, it should go to the doc in https://github.com/kubevirt/kubevirt/pull/11286 . Let us discuss my misunderstanding there.